riot.py

```diff
@@ -0,0 +1,480 @@
+
+    [(X, [X0, X1, ...]), (Y, [Y0, Y1, ...)] ->
+      ((X, X0), (Y, Y0)), ((X, X0), (Y, Y1)), ((X, X1), (Y, Y0)), ((X, X1), (Y, Y1))
+    """
+    all_vals = []
+
+    for name, vals in specs:
+        all_vals.append([(name, val) for val in vals])
+
+    all_vals = itertools.product(*all_vals)
+    return all_vals
+
+
+def case_iter(case):
+    # We could itertools.product here again, but I think this is clearer.
+    for env_cfg in expand_specs(case.env) if "env" in case else [[]]:
+        for py in case.pys:
+            for pkg_cfg in expand_specs(case.pkgs):
+                yield env_cfg, py, pkg_cfg
+
+
+def cases_iter(cases):
+    """Iterator over all case instances of a suite.
+
+    Yields the dependencies unique to the instance of the suite.
+    """
+    for case in cases:
+        for env_cfg, py, pkg_cfg in case_iter(case):
+            yield case, env_cfg, py, pkg_cfg
+
+
+def suites_iter(suites, pattern, py=None):
+    """Iterator over an iterable of suites.
+
+    :param py: An optional python version to match against.
+    :param pattern: An optional pattern to match suite names against.
+    """
+    for suite in suites:
+        if not pattern.match(suite.name):
+            logger.debug("Skipping suite '%s' due to mismatch.", suite.name)
+            continue
+
+        for case, env, spy, pkgs in cases_iter(suite.cases):
+            if py and spy != py:
+                continue
+            yield CaseInstance(suite=suite, case=case, env=env, py=spy, pkgs=pkgs)
+
+
+def run_suites(
+    suites, pattern, skip_base_install=False, recreate_venvs=False, out=sys.stdout, pass_env=False,
+):
+    """Runs the command for each case in `suites` in a unique virtual
+    environment.
+    """
+    results = []
+
+    # Generate the base virtual envs required for the test suites.
+    # TODO: errors currently go unhandled
+    generate_base_venvs(suites, pattern, recreate=recreate_venvs, skip_deps=skip_base_install)
+
+    for case in suites_iter(suites, pattern=pattern):
+        base_venv = get_base_venv_path(case.py)
+
+        # Resolve the packages required for this case.
+        pkgs: t.Dict[str, str] = {name: version for name, version in case.pkgs if version is not None}
+
+        # Strip special characters for the venv directory name.
+        venv = "_".join(["%s%s" % (n, rmchars("<=>.,", v)) for n, v in pkgs.items()])
+        venv = "%s_%s" % (base_venv, venv)
+        pkg_str = " ".join(["'%s'" % get_pep_dep(lib, version) for lib, version in pkgs.items()])
+        # Case result which will contain metadata about the test execution.
+        result = AttrDict(case=case, venv=venv, pkgstr=pkg_str)
+
+        try:
+            # Copy the base venv to use for this case.
+            logger.info("Copying base virtualenv '%s' into case virtual env '%s'.", base_venv, venv)
+            try:
+                run_cmd(["cp", "-r", base_venv, venv], stdout=subprocess.PIPE)
```

Comment:
I'd use `shutil.copytree` instead?

Reply:
Originally had it but it won't allow you to overwrite a directory if it already exists.